Actuators Document Improvement (#1858)
PX4/PX4-user_guide: en/config/actuators.md

Tilt servos can provide torque on one or more axes, which may be used to yaw or pitch the vehicle:
- Yaw is commonly set in this way, though motors are often used instead on vehicles with four or more motors.
- Pitch is more commonly controlled using differential motors thrust.
  Control using tilt servos is useful on airframes that can't use differential thrust, such as a [Bicopter](https://www.youtube.com/watch?v=hfss7nCN40A).

Whether this feature is used is configured in the `Use for control` setting.

## Actuator Outputs

The _Actuator Outputs_ section is used to assign motors, control surface servos, and other actuators used by the particular frame to the physical outputs on the flight controller.

Separate tabs are displayed for each output bus _that is present_ on the connected flight controller: PWM AUX, PWM MAIN, UAVCAN.
Motors and actuators (which are referred to as "functions") can be assigned to any physical output on any of the available busses.

![Actuator Outputs - Multicopter diagram](../../assets/config/actuators/qgc_actuators_mc_outputs.png)

:::note
PWM AUX outputs are preferred over the PWM MAIN outputs for controlling motors (they have lower latency).
:::

> PWM outputs are grouped according to the hardware groups of the autopilot.

Reviewer: Yes I think that's right.

### Control Surfaces Geometry

The control surfaces section of the geometry panel lets you set the number of control surfaces that are present on the vehicle.
You can then specify the type of each individual surface, and configure the appropriate scaling values for the Roll scale, Yaw scale, Pitch scale, Trim and Slew rate (advanced).

An "example" control surface section for a vehicle with two ailerons is shown below.
Note that ailerons only affect roll, so the pitch and yaw fields are disabled.

![Control Surface Setup Example](../../assets/config/actuators/control_surfaces_geometry.png)

The fields are:

- `Control Surfaces`: The number of control surfaces (set this first!)
- `Type`: The type of each control surface: `LeftAileron`, `RightAileron`, `Elevator`, `Rudder`, `Left Elevon`, `Right Elevon`, `Left V-Tail`, `Right V-Tail`, `Left Flap`, `Right Flap`, `Airbrakes`, `Custom`.
- `Roll scale`: Roll scale (normalised: -1 to 1)
- `Pitch scale`: Pitch scale (normalised: -1 to 1)
- `Yaw scale`: Yaw scale (normalised: -1 to 1)

> - `Trim`: ?

Reviewer 2: @sfuhrer Is this OK? Looking at it, my thought was "how do you work out what it is"? Is it effectively trial and error - i.e. in the actuator panel you set the actuator to the minimum value. If it isn't centred you change this value until it is?
```suggestion
- `Trim`: An offset added to the actuator so that it is centered without input (normally).
```

Reviewer: Yes that looks fine for now. And yes, that's what Trim is doing, yes I do it normally just by trial and error, but one could also check the convention for surface deflection. I don't think the "(normally)" is required.

> - `Skew`: ?

Reviewer 2: @bkueng Silvan says of skew "I don't know, where is that used?". It's an advanced option for actuators - but I haven't seen it set in any screens. Can you tell me the intent or some frame in which it is needed?

Reviewer: I couldn't find any reference in code nor in QGC to "skew". Did you by chance misread and meant "Slew rate"?
![image](https://user-images.githubusercontent.com/26798987/171568838-87f4efdb-5f25-4945-83c4-6017f52d54b0.png)

That would be the same as we had on the legacy mixing system: https://docs.px4.io/v1.12/en/concept/mixing.html#summing-mixer
"A minimal actuator traversal time limit can also be specified in the output scaler (inverse of a slew rate)."

Reviewer 2: I'm an idiot. Let us never speak of this again.

Tilt servos can provide torque on one or more axes, which may be used to yaw or pitch the vehicle:
- Yaw is commonly set in this way, though motors are often used instead on vehicles with four or more motors.
- Pitch is more commonly controlled using differential motors thrust.
  Control using tilt servos is useful on airframes that can't use differential thrust, such as a [Bicopter](https://www.youtube.com/watch?v=hfss7nCN40A).

Whether this feature is used is configured in the `Use for control` setting.

## Actuator Outputs

The _Actuator Outputs_ section is used to assign motors, control surface servos, and other actuators used by the particular frame to the physical outputs on the flight controller.

Separate tabs are displayed for each output bus _that is present_ on the connected flight controller: PWM AUX, PWM MAIN, UAVCAN.
Motors and actuators (which are referred to as "functions") can be assigned to any physical output on any of the available busses.

![Actuator Outputs - Multicopter diagram](../../assets/config/actuators/qgc_actuators_mc_outputs.png)

:::note
PWM AUX outputs are preferred over the PWM MAIN outputs for controlling motors (they have lower latency).
:::

> PWM outputs are grouped according to the hardware groups of the autopilot.
> Each group allows to configure the PWM rate or DShot/Oneshot (if supported).

Reviewer 2: ```suggestion
PWM outputs are grouped based on the protocols that they support.
In other words, the outputs that support DShot/Oneshot/PWM would be grouped together, separate from those that don't support Dshot (say).

```

The easiest way to try this out in simulation is to use any of the following gazebo `make` targets, which have control allocation pre-enabled:
```
make px4_sitl gazebo_iris_ctrlalloc
make px4_sitl gazebo_typhoon_h480_ctrlalloc
```
:::

Open the view in *QGroundControl* here: **"Q" (app menu) > Vehicle Setup > Actuators** (tab).

## Overview

The view has three sections:
- [Geometry](#geometry): Configure the geometry for the [selected airframe](../config/airframe.md).
  This includes number, position and properties of [motors](#motor-geometry) and also the number and properties of [control surfaces](#control-surfaces-geometry) and [motor tilt servos](#motor-tilt-servo-geometry).
- [Actuator Outputs](#actuator-outputs): Assign motors, control surfaces, and other actuators to specific output.
- [Actuator Testing](#actuator-testing): Test that motors and actuators move in the direction / speed as expected.

A quadcopter might have an setup screen similar to the one shown below.
This defines a 4 motor copter with X-geometry, and maps each of the motors to DShot ESC.

> It also maps PWM400 AUX outputs for controlling a parachute and landing gear as well.

Reviewer 2: Nope. You can do "It maps .... as well", or have it as it was. But the addition at the end is redundant.

```suggestion
It also maps PWM400 AUX outputs for controlling a parachute and landing gear.
```

![Actuators MC (QGC)](../../assets/config/actuators/qgc_actuators_mc_aux.png)

:::note
Only the most common settings are displayed by default.
Select the **Advanced** checkbox in the top right corner to display all settings.
:::

## Geometry

The geometry section is used to set any configurable geometry-related parameters for the selected [airframe](../config/airframe.md).
This includes the number and position of [motors](#motor-geometry), and the number, function, and properties of [control surfaces](#control-surfaces-geometry).
For VTOL tiltrotor vehicles, it will also include the number and properties of [tilt servos](#motor-tilt-servo-geometry)

:::note

> The UI shows the configured motor positions, customized to your vehicle (Currently only for Multicopter)

Reviewer 2: ```suggestion
The UI is customised for the selected airframe:
- Only configurable fields for the selected airframe type are displayed (i.e. if motor geometry isn't configurable, the fields won't be visible).
- The motor position diagram is currently only displayed for multicopter frames.
```

:::

### Motor Geometry

The motor geometry section lets you set the number of motors, the relative position, and other properties of each motor.

Most motor properties apply to all frames.
A few properties apply to specific frames.

> For example, `Tilted-by` and `axis` are only shown & relevant for [Tiltrotor VTOL](#motor-geometry-vtol-tiltrotor) and [Standard VTOL](#motor-geometry-standard-vtol) vehicles, respectively.

Reviewer 2: ```suggestion
For example, `Tilted-by` and `axis` are only relevant for [Tiltrotor VTOL](#motor-geometry-vtol-tiltrotor) and [Standard VTOL](#motor-geometry-standard-vtol) vehicles, respectively.
```

Note that you will need to also ensure that the ESC associated with bidirectional motors is configured appropriately (e.g. 3D mode enabled for DShot ESCs, which can be achieved via [DShot commands](../peripherals/dshot.md#commands)).

### Control Surfaces Geometry

The control surfaces section of the geometry panel lets you set the number of control surfaces that are present on the vehicle.
You can then specify the type of each individual surface, and configure the appropriate scaling values for the Roll scale, Yaw scale, Pitch scale, Trim and Slew rate (advanced).

An "example" control surface section for a vehicle with two ailerons is shown below.
Note that ailerons only affect roll, so the pitch and yaw fields are disabled.

![Control Surface Setup Example](../../assets/config/actuators/control_surfaces_geometry.png)

The fields are:

- `Control Surfaces`: The number of control surfaces (set this first!)
- `Type`: The type of each control surface: `LeftAileron`, `RightAileron`, `Elevator`, `Rudder`, `Left Elevon`, `Right Elevon`, `Left V-Tail`, `Right V-Tail`, `Left Flap`, `Right Flap`, `Airbrakes`, `Custom`.

> - `Roll scale`: Roll scale (normalised: -1 to 1)
> - `Pitch scale`: Pitch scale (normalised: -1 to 1)
> - `Yaw scale`: Yaw scale (normalised: -1 to 1)

Reviewer 2: @sfuhrer Thanks for your [response above](https://github.com/PX4/PX4-user_guide/pull/1858/files#r881534338) on these.  To improve this I added a section with your information about the scaling below - https://github.com/PX4/PX4-user_guide/pull/1858/files#r882271662

Then I link to this here. Is this OK?

```suggestion
- `Roll scale`: Effectiveness of actuator around roll axis (normalised: -1 to 1). [Generally you should use the default actuator value](#actuator-roll-pitch-and-yaw-scaling).
- `Pitch scale`: : Effectiveness of actuator around pitch axis (normalised: -1 to 1).  [Generally you should use the default actuator value](#actuator-roll-pitch-and-yaw-scaling).
- `Yaw scale`: Effectiveness of actuator around yaw axis (normalised: -1 to 1).  [Generally you should use the default actuator value](#actuator-roll-pitch-and-yaw-scaling).
```

- `Trim`: ?
- `Skew`: ?
- `Lock control surfaces in hover`:
  - `Enabled`: Most vehicles do not use control surfaces in hover. Use this setting to lock them so that they don't affect vehicle dynamics.
 - `Disabled`: Set this for vehicles that use control surfaces in hover, such as the duo tailsitter (which uses elevons for pitch and yaw control). It should also be set for vehicles that use control surfaces to provide additional stabilization in hover mode when moving at speed or in high winds.

> #### Actuator Roll, Pitch, and Yaw Scaling

Reviewer 2: @sfuhrer FYI only I put the detail on scaling you provided here.

> - `Slew Rate`: Minimum time allowed for the motor/servo signal to pass through the full output range.
>    - It is intended for actuators that may be damaged if they move too fast — like the tilting actuators on a tiltrotor VTOL vehicle. It can be used to limit the rate of change of an actuator (if this is not specified then no rate limit is applied).
>    - For example, a setting of 2.0 means that the motor/servo signal can only go from 0 to 1 in minimum x seconds (in case of reversible motors, the range is -1 to 1).

Reviewer 2: ```suggestion
- `Slew Rate`: Minimum time allowed for the motor/servo signal to pass through the full output range, in seconds.
   - It is intended for actuators that may be damaged if they move too fast — like the tilting actuators on a tiltrotor VTOL vehicle.
     The setting limits the rate of change of an actuator (if not specified then no rate limit is applied).
   - For example, a setting of 2.0 means that the motor/servo should not be commanded to move from 0 to 1 at a rate that completes the operation in less than 2 seconds (in case of reversible motors, the range is -1 to 1).
```

Author: I think the `will not be commanded` fits better than `should not be commanded`, since control allocation module is the one in charge, and hence `should` implies that user has to take that into account somehow :thinking:

Reviewer 2: @junwoo091400 Good point, I patched this in https://github.com/PX4/PX4-user_guide/commit/20a4ae785c91976983a3d75d2692f410da67208d

1. Repeat for all actuators

### Motor Configuration

The motor configuration sets output values such that motors:
- don't spin when disarmed (at the `disarmed` PWM output value)
- barely spin at the `minimum` PWM output value
- give **positive thrust** in the expected direction

For each motor:

1. Pull the motor slider down so that it snaps to the bottom.
   In this position the motor is set to the outputs `Disarmed` value.
   - Verify that the motor doesn't spin in this position.
   - If the motor spins, reduce the corresponding PWM `Disarmed` value in the [Actuator Outputs](#actuator-outputs) section to below the level at which it still spins.
2. Slowly move the slider up until it snaps to the _minimum_ position.
   In this position the motor is set to the outputs `Minimum` value.  
   - Verify that the motor is spinning very slowly in this position.
   - If the motor is not spinning, or spinning too fast you will need to adjust the corresponding PWM `Disarmed` value in the [Actuator Outputs](#actuator-outputs) such that the motors barely spin.
   
     ![PWM Minimum Output](../../assets/config/actuators/pwm_minimum_output.png)
     :::note

>      For DShot output, this is not required <!-- any, or just the minimum check? -->

Reviewer 2: Is it just this step or any of this configuration not required for DShot?

     :::
3. Increase the slider value to a level where you can verify that the motor is spinning in the correct direction and that it would give a positive thrust in the expected direction.
   - The expected thrust direction can vary by vehicle type, for example in multicopters the thrust should always point upwards, while in a fixed wing vehicle the thrust will push the vehicle forwards.
   - For VTOL, thrust should point upwards when the Tilt Servo is at 0 degrees as defined the [Tilt Servo Convention](#motor-tilt-servos-conceptual).
     Testing of the [Tilt Servo](#tilt-servo-testing) is covered below as well.
   - If thrust is in the wrong direction, you may need to [reverse the motors](#reversing-motors).
   

### Control Surface Setup

The control surface configuaration sets output values such that the surfaces:

1. have the correct idle position when disarmed.
1. move in the directions as defined in the [Control Surface Convention](#control-surface-direction-conceptual).

![Control Surface Disarmed 1500 Setting](../../assets/config/actuators/control_surface_disarmed_1500.png)

For each control surface:

> 1. Set the `Disarmed` value for Control Surface outputs to `1500` (note, this assumes we're using a PWM actuator where 1500 is the middle of the range)

Reviewer 2: @junwoo091400 I gave it a big edit of the output section. Quite happy with the structure now - following your idea that they are entwined with testing. Also separated out an output assignment section that is "manual".

Just this section and the one after to go. My problem is that, while I get the point, I don't think it is generic.

For a start, are you saying that all control surfaces are controlled using PWM values of range 1000 to 2000. Reading this I think so. Further, I think that you set the value always to 1500. Is that correct, or would you adjust it if it was in the wrong position? Or would you adjust the trim value in the geometry?

Author: The 'Disarmed' value and 'Trim' values are decoupled and can be quite confusing.

'Disarmed' is which command the Servo will be commanded when the vehicle is disarmed. And it's absolute value / precision is not so relevant, as you just want to have the control surfaces in a reasonable position (and not in min/max position).

'Trim' is relevant for when the control surface is actively being controlled. As far as I know, if the trim = 0, command will be based on the average of the minimum / maximum PWM range. So the 'disarmed' isn't related at all.

   - Unlike motors, when `disarmed` we usually control surfaces to be in the middle position of their range.
   - Adjust the value until the control is positioned correctly.

> 2. Set the Output mode to `PWM 50Hz`, instead of default PWM 400 Hz when using a PWM output for Control Surfaces

Reviewer 2: What? Now you're saying I change my mode, and now I use PWM50Hz rather that PWM400Hz? This makes no sense.

What I "think" you mean is that you generally don't need as high speed control for actuators, so you're recommending a slow protocol to get more reliability. Is that right?

So I think you need to start by saying all of that BEFORE you have any of these steps. And then describe the rest of the setting relative to just one PWM setting.

Author: Gotcha, and yes it's to prevent the Servo from making weird noises when it's commanded with PWM at 400 Hz. And because the default configuration is 400 Hz, I added the command there :thinking: 

Since it applies to the whole commands, I agree it's better to put it in the beginning.

   - If your servo is creating weird noise, it is most likely because of the "PWM 400Hz" output, which is usually not accepted by commercial Servos.
   - In case your Servo supports 100, 200 or 400 Hz command, you can of course set the output mode to the maximum PWM frequency it supports!
3. Move Control Surface slider upwards (positive command) and verify that it moves in the direction defined in the [Control Surface Convention](#control-surface-direction-conceptual).
   1. If the control surface moves in the opposite direction, click on the 'Rev Range' checkbox to reverse the range
4. Move the slider to the middle and check if the Control Surfaces are aligned in the neutral position of the wing

    ![Control Surface Trimming](../../assets/config/actuators/control_surface_trim.png)

>    1. If it is not aligned, you can set the **Trim** value for each control surface.

Reviewer 2: Now we're talking about trimming. I'm even more confused

Author: Not sure why you are confused. This is a procedure of setting up a control surface that any user would go through :thinking:

   2. After setting a new trim, manually click and place the slider for the control surface you are testing in the middle position, to check if it is in the neutral position.
   3. Note that as mentioned above as a safety feature of slider, unless you actually touch the slider, the actuator command doesn't get sent.
      Therefore **you must manually click and put the slider in the middle position to test the trim**
   4. Or, an easier way to test without using the slider would be to set the [`COM_PREARM_MODE`](../advanced_config/parameter_reference.md#COM_PREARM_MODE) parameter to `Always`.
      1. Which will enable the control of Servos even when the vehicle is disarmed, and will constantly be applying the Trim setting to the Control Surfaces
      2. Here you can try setting different values for the Trim and check the alignment, and then settle on the value you are happy with.

### Tilt Servo Setup

> ![Tilt Servo Setup](../../assets/config/actuators/tilt_servo_setup.png)

Reviewer 2: Probably the same problems as above.

Selecting these requires a reboot before they are applied.

You should assign functions to the outputs that match your physical wiring of motors and servos, and use the [Actuator Testing](#actuator-testing) section described below to determine appropriate output parameter values.
These steps are covered in [Output Assignment and Configuration](#output-assignment-and-configuration).

## Actuator Testing

The _Actuator Testing_ section in lower-right corner provides sliders that can be used to test (and determine) actuator and motor settings. 
A slider is provided for each output defined in the [Actuator Outputs](#actuator-outputs) section.
The slider example below shows the section for a typical VTOL Tiltrotor airframe.

![Actuator Testing Slider](../../assets/config/actuators/vtol_tiltrotor_sliders_example.png)

The section has an **Enable Sliders** switch that must be toggled before sliders can be used.
The sliders can power the motors/servos across their full range of motion, and "snap" to the disarmed and minimum positions.

:::note
After you toggle the **Enable sliders** switch, actuators/motors won't do anything until the corresponding slider is _moved_.

> This is a safety feature to prevent sudden motor movements after switch is enabled.

Reviewer 2: I'm going to leave "to prevent sudden motor movements after switch is enabled" but I think it is "redundant". 

Why? Because the less people have to read the better. It is necessary to know that the actuators won't move unless the sliders move in this case. It is "good" to know that it is intentional behaviour. The extra detail of to prevent sudden motor movements is easily inferred.

It is "correct" though.

Author: I agree! It's redundant :disappointed:, now that I reflect back on it! I just remembered Beat telling me the reason for this safety feature, and I thought it wouldn't be so easily inferrable, but I guess it is :wink:

:::

Sliders can be used to verify the following:

1. Actuators (Motors, Control surfaces, etc.) are assigned to the expected output.
1. Motors don't spin when at the `disarmed` PWM output value
1. Motors barely spin at the `minimum` PWM output value.
1. Motors give **positive thrust** in the expected direction
1. Control Surfaces are in the correct idle position for `disarmed` output value
1. Control Surfaces move in the direction as defined in the [Control Surface Convention](#control-surface-direction-conceptual)
1. Motor Tilt Servos are in the correct idle position for `disarmed` output value
1. Motor Tilt Servos move in the direction as defined in the [Tilt Servo Convention](#motor-tilt-servos-conceptual)

## Output Assignment and Configuration

> In the _Actuator Outputs_ section on the right panel, you can assign and configure outputs, optionally using the [Actuator Testing](#actuator-testing) sliders:

Reviewer 2: ```suggestion
Outputs are assigned to functions and configured in the [Actuator Outputs](#actuator-outputs) section, while the  [Actuator Testing](#actuator-testing) sliders are commonly used to determine appropriate configuration values to enter.
```

Author: Actuator Testing also is used for determining control surface movement direction, trim and so on, so maybe this change makes it too specific to one aspect of Actuator Testing?

- MC vehicles that have connected motors to PWM outputs can use the [Identify & Assign Motors](#multicopter-pwm-identify-assign-motors) button to perform motor assignment "semi-automatically".

> - Motor output assignment can be done/checked using sliders (see [Output Assignment (Manual)](#output-assignment-manual)).
> - Actuator output assignment can be done/checked using sliders (see [Output Assignment (Manual)](#output-assignment-manual)). 

Reviewer 2: Merge these two:
```suggestion
- Output assignment (of motors and actuators) can be done/checked using sliders (see [Output Assignment (Manual)](#output-assignment-manual)).
```

You can use the **Identify & Assign Motors** button to assign motors to PWM outputs using a semi-automated process.

:::note
This is the easiest way to assign motors, but is currently only supported for motors on **multicopter vehicles** that are connected to PWM outputs (UAVCAN outputs and other frame types do not support this feature).
On other frames you can follow the instructions in [Output Assignment (Manual)](#output-assignment-manual).
:::

:::warning
Remove the propellers from motors before assigning outputs or any testing.
:::

![Identify motor button](../../assets/config/actuators/identify_motors_button.png)

When you click the button, QGC sends a command to a motor, causing it to spin.
To assign that motor to an output you simply select the corresponding motor displayed in the screen.
QGC will then spin the next motor for you to assign, and so on.

Instructions:
1. Setup the motor geometry to match the motors on your frame.
1. Select the PWM tab where you want to assign the motors.
1. Click the **Identify & Assign Motors** button.
1. One motor will start spinning (click **Spin Motor Again** if it stops spinning too quickly to note.)

>    Click on the corresponding motor in the geometry section.

Reviewer 2: Changing this back. Click and select are synonyms for a button, but click does not work on a touch screen.

```suggestion
   Select the corresponding motor in the geometry section.
```

Author: Haven't thought about that!

:::note
You will almost certainly need to change the pulse rate from the default of 400Hz because support is rare (if not supported the servo will usually make an "odd" noise).
If you're using PWM servos, PWM50 is far more common.
If a high rate servo is _really_ needed, DShot offers better value.
:::

For each of the control surfaces:

1. Set the `Disarmed` value so that the surfaces will stay at neutral position when disarmed.
   This is usually around `1500` for PWM servos.
2. Move the slider for the surface upwards (positive command) and verify that it moves in the direction defined in the [Control Surface Convention](#control-surface-direction-conceptual).
   - If the control surface moves in the opposite direction, click on the `Rev Range` checkbox to reverse the range.
3. Move the slider again to the middle and check if the Control Surfaces are aligned in the neutral position of the wing
   - If it is not aligned, you can set the **Trim** value for the control surface.
     :::note
	 This is done in the `Trim` setting of the Geometry panel, usually by "trial and error".
	 ![Control Surface Trimming](../../assets/config/actuators/control_surface_trim.png)

> 	 :::

Reviewer 2: ```suggestion
     :::
```

## Overview

The view has three sections:
- [Geometry](#geometry): Set any configurable geometry-related parameters for the [selected airframe](../config/airframe.md).
  This includes the number and relative position of motors, and the number, function, and properties of [control surfaces](#control-surfaces-geometry) (ailerons, rudders, etc.).
- [Actuator Outputs](): Assign motors, control surfaces, and other actuators to specific flight controller outputs.
- [Actuator Testing](): Test that motors and actuators are mapped to the correct outputs, and move as expected.

A quadcopter might have an setup screen similar to the one shown below.
This defines a 4 motor copter with X-geometry, and maps each of the motors to DShot ESC.
It also maps PWM400 AUX outputs for controlling a parachute and landing gear.

![Actuators MC (QGC)](../../assets/config/actuators/qgc_actuators_mc_aux.png)

Note that only the most common settings are displayed by default.
Select the **Advanced** checkbox in the top right corner to display all settings.

## Geometry

The geometry section is used to set any configurable geometry-related parameters for the selected [airframe](../config/airframe.md).

> This includes the number and position of [motors](#motor-geometry), and the number, function, and properties of [control surfaces](#control-surfaces-geometry).

Author: We could omit the 'function' part of the Motors part, I am not sure what you mean by 'function'?

For VTOL tiltrotor vehicles, it will also include the number and properties of [tilt servos](#motor-tilt-servo-geometry)

The UI is customised for the selected airframe.
Only motors and/or control surfaces that are relevant to the airframe type are displayed or available to assign.
If there are no configurable geometry options for a particular frame, this section could conceivably be empty.

:::note
Not all aspects of a geometry are necessarily configurable through this interface.
Factors that cannot be configured are hard coded for the current airframe. 

> :::

Author: What do you mean by this? Any examples?

Reviewer 2: We know for a fact that you can't use a multicopter definition with a vtol, so there are some aspects of the geometry which are not captured. I assume some of them are irrelevant or can be tuned away. 

Or I could just be wrong. @bkueng ?

### Motor Geometry

The motor geometry section lets you set the number motors, and the relative position, function, and other properties of each motor.

> The motor properties are similar for all frames, but there are some exceptions.

Author: What kind of 'exceptions' are there?

Reviewer 2: The ones that have different behaviours, outlined in the following sections. So for example, tiltotors additionally set the associated tilt servo, and standard quads have a direction applied to the motors.

Author: Aha gotcha 🙆‍♂️

#### Motor Geometry: Multicopter

The screenshot below shows the geometry screen for a multicopter frame with (right) and without (left) advanced settings.
Multicopters don't have any control surfaces, so none are shown.

![Geometry MC (QGC)](../../assets/config/actuators/qgc_actuators_mc_geometry.png)

The **Motors** setting lets you choose the number of motors.

For each multicopter motor you can then set:
- `Position X`: [X-position](#motor-position-coordinate-system), in metres.
- `Position Y`: [Y-position](#motor-position-coordinate-system), in metres.
- `Position Z`: [Z-position](#motor-position-coordinate-system), in metres.
- `Direction CCW`: Checkbox to indicate motor spins counter-clockwise (uncheck for clockwise).
- `Bidirectional`: Checkbox to indicate motor is [bidirectional](#bidirectional-motors) 
- `Slew Rate`: ?

The X, Y, Z positions are in [FRD relative to the centre of gravity](#motor-position-coordinate-system).

#### Motor Geometry: VTOL Tailsitter

The default tailsitter motor geometry is shown below.
Motors have the same configuration fields as for the [multicopter geometry](#motor-geometry-multicopter).

> ![Geometry motor: tailsitter vtol](../../assets/config/actuators/qgc_geometry_tailsitter_motors.png)

Author: I think you are referring to the Quad Tailsitter (https://docs.px4.io/master/en/airframes/airframe_reference.html#vtol-quad-tailsitter). But there also is the Duo Tailsitter (https://docs.px4.io/master/en/airframes/airframe_reference.html#vtol-duo-tailsitter), which is what I first thought of when I read the name "VTOL Tailsitter".

It would be better to clarify that this is a 'Quad' Tailsitter!

#### Motor Geometry: VTOL Tiltrotor

The default VTOL tiltrotor motor geometry is shown below.

> ![Geometry motor: tiltrotor vtol](../../assets/config/actuators/qgc_geometry_tiltrotor_motors.png)

Author: This doesn't really match with the Tiltrotor VTOL configuration (with 3 motors). I will replace it with the actual screenshot from the Converence setup :+1:

Reviewer 2: Sure!

- `Axis`: One of `Upwards`, `Downwards`, `Forwards`, `Backwards`, `Leftwards`, `Rightwards`, `Custom`
  - If `Custom` is selected, then the UI displays three additional fields for setting the motor orientation.

#### Motor Geometry: Other Vehicles

Other vehicle types will define an appropriate motor geometry for their frame type.
Once again these motors will generally have the same kinds of properties as shown above.

For example, a fixed-wing vehicle may just have a single pusher moter, while a rover with differential steering will have a motor for throttle and for steering.

#### Motor Position Coordinate System

The coordinate system for motor positions is FRD (in body frame), where the X axis points forward, the Y axis to the right and the Z axis down.
The origin of the system is the vehicle centre-of-gravity (COG) (**not** the autopilot location).

![Actuators CG reference diagram](../../assets/config/actuators/quadcopter_actuators_cg_reference.png)

#### Bidirectional Motors

Some vehicles may use bidirectional motors (i.e. motor spins in direction 1 for lower output range and in direction 2 for the upper half).

> For example, ground vehicles that want to move forwards and backwards, or VTOL vehicles that have pusher motors that go in either direction.

Author: What is this 'VTOL vehicle with pusher that can go in either direction' use case? Is this a real use case?

Reviewer 2: Yes, according to Beat. I guess the pusher can act as an airbrake in that case.

To get a feeling for the motor layout for VTOL and other vehicles see the [Airframe Reference](../airframes/airframe_reference.md)

The motor geometry sections for a number of different frames are shown below.

#### Motor Geometry: Multicopter

The screenshot below shows the geometry screen for a multicopter frame with (right) and without (left) advanced settings.
Multicopters don't have any control surfaces, so none are shown.

![Geometry MC (QGC)](../../assets/config/actuators/qgc_actuators_mc_geometry.png)

The **Motors** setting lets you choose the number of motors.

For each multicopter motor you can then set:
- `Position X`: [X-position](#motor-position-coordinate-system), in metres.
- `Position Y`: [Y-position](#motor-position-coordinate-system), in metres.
- `Position Z`: [Z-position](#motor-position-coordinate-system), in metres.
- `Direction CCW`: Checkbox to indicate motor spins counter-clockwise (uncheck for clockwise).
- `Bidirectional`: Checkbox to indicate motor is [bidirectional](#bidirectional-motors) 

> - `Slew Rate`: ?

Reviewer 2: What is slew rate?

> The X, Y, Z positions are in [FRD relative to the centre of gravity](#motor-position-coordinate-system).

Reviewer 2: Here I forward link to the position information

Motors have most of the same configuration fields as for the [multicopter geometry](#motor-geometry-multicopter).
There is an additional field to indicate which servo is used to tilt the motor:

- `Tilted by`: The associated servo used for tilting the motor.
  The properties of this servo are defined in the [Motor Tilt Servo Geometry](#motor-tilt-servo-geometry).

#### Motor Geometry: Standard VTOL

The default standard VTOL ("quadplane") motor geometry is shown below.

![Geometry motor: standard vtol](../../assets/config/actuators/qgc_geometry_standard_vtol_motors.png)

Motors have most of the same configuration fields as for the [multicopter geometry](#motor-geometry-multicopter).
There is an additional field to indicate the direction in which the motor moves the vehicle (for a standard VTOL, the hover motors are usually set "upwards" and the pusher motor is set to "forwards").

- `Axis`: One of `Upwards`, `Downwards`, `Forwards`, `Backwards`, `Leftwards`, `Rightwards`, `Custom`
  - If `Custom` is selected, then the UI displays three additional fields for setting the motor orientation.

#### Motor Geometry: Other Vehicles

> Other vehicle types will define an appropriate motor geometry for their frame type.

Reviewer 2: This is just a kind of placeholder. If we want we could add boat, etc here too. We might in future, or split out instructions for the different vehicle types.

The origin of the system is the vehicle centre-of-gravity (COG) (**not** the autopilot location).

![Actuators CG reference diagram](../../assets/config/actuators/quadcopter_actuators_cg_reference.png)

#### Bidirectional Motors

Some vehicles may use bidirectional motors (i.e. motor spins in direction 1 for lower output range and in direction 2 for the upper half).
For example, ground vehicles that want to move forwards and backwards, or VTOL vehicles that have pusher motors that go in either direction.

If bidiectional motors are used, make sure to select the **Reversible** checkbox for those motors (the checkbox is displayed as an "advanced" option).

![Reversible](../../assets/config/actuators/qgc_geometry_reversible.png)

Note that you will need to also ensure that the ESC associated with bidirectional motors is configured appropriately (e.g. 3D mode enabled for DShot ESCs, which can be achieved via [DShot commands](../peripherals/dshot.md#commands)).

<!-- For example, the roll scale, pitch scale, slew rate and trim can be set (where appropriate) for a control surface, but the positions of these surfaces to each other and to motors cannot be specified. -->

> ### Control Surfaces Geometry

Reviewer 2: This bit below is very hard to read. @junwoo091400 I was thinking of making it more "real" like the way I did for multicopters. So none of this "conceptual" heading stuff - use a real example and then extend it. 

As it is now all the information is there, but I think people will find it hard to map to what they are actually doing in the section.

So actually I'd start with a standard vtol (say) and show its actuators and list all the possible options. I might then link down to explain the specific types.

I will look at this tomorrow.

* Example : To pitch in positive direction (pitching skywards), 'Both Ailerons (Right & Left)' need a positive deflection.

'Roll RIGHT' control input should command positive deflection to right aileron and negative deflection to left aileron

* Example : To roll in positive direction (banking to the right), the 'Right Aileron' needs a positive deflection.

#### Vertical Control Surface (e.g. Rudder)

> Rightwards movement equals Positive Deflection

'Yaw RIGHT' control input should command positive deflection.

* Example : To yaw in positive direction (clockwise rotation), conventional rudder airplane needs a positive deflection.

#### Mixed Control Surface (e.g. V-Tail)

> Upwards movement equals Positive deflection

'Pitch UP' control input should command positive deflection to both V-tails

* Example : To pitch in positive direction (pitching upwards),  'Both V-tail surfaces' need a positive deflection.

> Rightwards movement equals Positive Deflection
'Yaw RIGHT' control input should command positive deflection to both V-tails.

* Example : Having a V or A shaped tail in the back, positive deflection in Rightwards direction will Yaw the vehicle to the right.

> ### Motor Tilt Servo Geometry

Reviewer 2: OK, so here I updated your Motor tilt bit into its own "geometry" section. It cross links back to the tilt motor section.

@bkueng Arguably, it might be good for this section in the geometry to appear just below the motor configuration, since they are closely linked.

This section not quite complete. In particular, I don't understand the reference frame, and how the vehicles tilt.

This includes the number and position of [motors](#motor-geometry), and the number, function, and properties of [control surfaces](#control-surfaces-geometry).
For VTOL tiltrotor vehicles, it will also include the number and properties of [tilt servos](#motor-tilt-servo-geometry)

The UI is customised for the selected airframe.
Only motors and/or control surfaces that are relevant to the airframe type are displayed or available to assign.

:::note
Not all aspects of a geometry are necessarily configurable through this interface.
Factors that cannot be configured are hard coded for the current airframe. 
:::

### Motor Geometry

The motor geometry section lets you set the number of motors, the relative position, and other properties of each motor. The motor properties are similar for all frames, but there are some exceptions.

Multicopter airframe provides a diagram in QGC showing the X, Y positions, which is useful for checking whether the motors configuration is correct. To get a feeling for the motor layout for VTOL and other vehicles see the [Airframe Reference](../airframes/airframe_reference.md)

> The motor geometry explanations for a number of different frames are described below.

Reviewer 2: FYI This says "you are describing the explanations" not "describing the frames". I will fix.

> #### Motor Position Coordinate System

Reviewer 2: I wasn't entirely convinced this should go first. But let's try it here.

The **origin is the vehicle's centre-of-gravity (COG)**, and **NOT** the autopilot location).

![Actuators CG reference diagram](../../assets/config/actuators/quadcopter_actuators_cg_reference.png)

#### Motor Geometry: Multicopter

The screenshot below shows the geometry screen for a multicopter frame with (right) and without (left) advanced settings.
Multicopters don't have any control surfaces, so none are shown.

![Geometry MC (QGC)](../../assets/config/actuators/qgc_actuators_mc_geometry_marked.png)

First, the **Motors** drop-down setting lets you choose the number of motors, which is 4 for the example above.

For each motor you can then set:
- `Position X`: [X-position](#motor-position-coordinate-system), in metres.
- `Position Y`: [Y-position](#motor-position-coordinate-system), in metres.
- `Position Z`: [Z-position](#motor-position-coordinate-system), in metres.
- (Advanced) `Direction CCW`: Checkbox to indicate motor spins counter-clockwise (uncheck for clockwise).
- (Advanced) `Bidirectional`: Checkbox to indicate motor is [bidirectional](#bidirectional-motors) 

> - (Advanced) `Slew Rate`: ?

Reviewer 2: What does this mean? Definition needed

The X, Y, Z positions are in [FRD coordinate frame, relative to the centre of gravity](#motor-position-coordinate-system).

#### Motor Geometry: VTOL Quadrotor Tailsitter

VTOL Quadrotor is a tailsitter with 4 motors like a quadcopter as shown below.

> ![VTOL Quadrotor Tailsitter](../../assets/airframes/types/VTOLQuadRotorTailSitter.svg)

Reviewer 2: These images are too big. Let's just link to other docs.

#### Motor Geometry: Other Vehicles

Other vehicle types will define an appropriate motor geometry for their frame type.
Once again these motors will generally have the same kinds of properties as shown above.

For example, a fixed-wing vehicle may just have a single pusher moter, while a rover with differential steering will have a motor for throttle and for steering.

#### Bidirectional Motors

Some vehicles may use bidirectional motors (i.e. motor supports spinning in both directions)
For example, ground vehicles that want to move forwards and backwards, or VTOL vehicles that have pusher motors that go in either direction.

If bidiectional motors are used, make sure to select the **Reversible** checkbox for those motors (the checkbox is displayed as an "advanced" option).

![Reversible](../../assets/config/actuators/qgc_geometry_reversible_marked.png)

Note that you will need to also ensure that the ESC associated with bidirectional motors is configured appropriately (e.g. 3D mode enabled for DShot ESCs, which can be achieved via [DShot commands](../peripherals/dshot.md#commands)).

> ### Control Surfaces Geometry

Reviewer 2: @bkueng @junwoo091400 I have updated this control surfaces geometry section. It needs more input from @bkueng now.

Why? All the stuff about how control surfaces work doesn't tell people how to set the values here. 

@bkueng around about line 163 we have this:
```
- `Roll scale`: Roll scale (normalised: -1 to 1)
- `Pitch scale`: Pitch scale (normalised: -1 to 1)
- `Yaw scale`: Yaw scale (normalised: -1 to 1)
- `Trim`: ?
- `Skew`: ?
```
What should we put for each of these? I.e. what do values in the range correspond to? What is the input, and what is the output if (say) the value is 0.7?

Is the "Control Surface Deflection Convention" useful for explaining the above? Left in place in case it is!

Reviewer 3: @sfuhrer I leave this up to you

Reviewer 2: @sfuhrer Can you comment on this whole section on control surfaces?
In particular, as I asked in ^^^
- What should we put for each of the fields? I.e. what do values in the range correspond to? What is the input, and what is the output if (say) the value is 0.7?
  - In other words, consider someone with a VTOL or FW frame that has an elevon (say). How do they decide what number to put in here?
- How do we describe trim? 
  - I "think" it it is like an offset to the default expected position. So you have a servo where you'd liked the lowest input value to be at the default position. However the unpowered servo is actually a bit less than this. So you're saying that 0 is actually at 0.1 say and that code should automatically account for this when setting values.
  - But I could be completely wrong, and that still doesn't tell me what the scale is. So same question. I have a vehicle, what is this trim, and how do I set it.
- Ditto for skew?

Reviewer: >What should we put for each of the fields? I.e. what do values in the range correspond to? What is the input, and what is the output if (say) the value is 0.7?

It's a normalized effectiveness of the actuator around the corresponding axis. The numbers that are entered there are directly put into the allocation matrix, that is then inverted to get from desired moments (normalized) to control signals. Increasing the scale will reduce the deflection of the control surfaces (as it gets inverted). 
This is very low level stuff that a normal user shouldn't care about, unless it is for the tuning of coupled control surfaces (like an elevon, that controls both pitch and roll). In that case we should mention what I wrote above (if you want more deflection of a certain actuator around a certain axis, you can reduce the scale). But really make sure that for most setups the default scales are all you need (they appear automatically when selecting the control surface type).

Trim: yes it's simply an offset that is added on this actuator, normally to have it centered without input.

>Ditto for skew?

I don't know, where is that used?

We can link the devsummit talk Beat and I will make on CA (once we have it), that will include a walk-through.

### Control Surfaces Geometry

The control surfaces section of the geometry panel lets you set the number of control surfaces that are present on the vehicle.
You can then specify the type of each individual surface, and configure the appropriate scaling values for the Roll scale, Yaw scale, Pitch scale, Trim and Slew rate (advanced).

An "example" control surface section for a vehicle with two ailerons is shown below.
Note that ailerons only affect roll, so the pitch and yaw fields are disabled.

![Control Surface Setup Example](../../assets/config/actuators/control_surfaces_geometry.png)

The fields are:

- `Control Surfaces`: The number of control surfaces (set this first!)
- `Type`: The type of each control surface: `LeftAileron`, `RightAileron`, `Elevator`, `Rudder`, `Left Elevon`, `Right Elevon`, `Left V-Tail`, `Right V-Tail`, `Left Flap`, `Right Flap`, `Airbrakes`, `Custom`.
- `Roll scale`: Roll scale (normalised: -1 to 1)
- `Pitch scale`: Pitch scale (normalised: -1 to 1)
- `Yaw scale`: Yaw scale (normalised: -1 to 1)
- `Trim`: ?
- `Skew`: ?

> - `Lock control surfaces in hover`: `Enabled` or `Disabled`.

Reviewer: Most VTOLs don't require the control surfaces in hover for control (exception: duo tailsitter that uses the elevons for pitch and yaw control). Sometimes you still want to enable them in hover already, as they can help the MC motors in stabilizing the vehicle in wind or when flying fast.

> - `Lock control surfaces in hover`: `Enabled` or `Disabled`.

Reviewer 2: Thanks @sfuhrer ! I guess if they aren't "required" then locking them prevents them moving around an potentially affecting vehicle control?

Is this OK?
```suggestion
- `Lock control surfaces in hover`:
  - `Enabled`: Most vehicles do not use control surfaces in hover. Use this setting to lock them so that they don't affect vehicle dynamics.
 - `Disabled`: Set this for vehicles that use control surfaces in hover, such as the duo tailsitter (which uses elevons for pitch and yaw control). It should also be set for vehicles that use control surfaces to provide additional stabilization in hover mode when moving at speed or in high winds.
```

> - `Lock control surfaces in hover`: `Enabled` or `Disabled`.

Reviewer: >Thanks @sfuhrer ! I guess if they aren't "required" then locking them prevents them moving around an potentially affecting vehicle control?

Yes correct, they usually do more harm then good. 

And yes, I like your proposal :+1:

You can then specify the type of each individual surface, and configure the appropriate scaling values for the Roll scale, Yaw scale, Pitch scale, Trim and Slew rate (advanced).

An "example" control surface section for a vehicle with two ailerons is shown below.
Note that ailerons only affect roll, so the pitch and yaw fields are disabled.

![Control Surface Setup Example](../../assets/config/actuators/control_surfaces_geometry.png)

The fields are:

- `Control Surfaces`: The number of control surfaces (set this first!)
- `Type`: The type of each control surface: `LeftAileron`, `RightAileron`, `Elevator`, `Rudder`, `Left Elevon`, `Right Elevon`, `Left V-Tail`, `Right V-Tail`, `Left Flap`, `Right Flap`, `Airbrakes`, `Custom`.
- `Roll scale`: Roll scale (normalised: -1 to 1)
- `Pitch scale`: Pitch scale (normalised: -1 to 1)
- `Yaw scale`: Yaw scale (normalised: -1 to 1)
- `Trim`: ?
- `Skew`: ?
- `Lock control surfaces in hover`: `Enabled` or `Disabled`.

The scaling values are normalised (-1 to 1), decoupling the outputs of PX4 controllers from the physical servo control signals.

> #### Control Surface Deflection Convention

Reviewer 2: @sfuhrer This section was originally added to help us describe the control surface values - as I ask here: https://github.com/PX4/PX4-user_guide/pull/1858/files#r881074960

I wasn't sure it was useful for that purpose so I left it in - i.e. to indicate the frame and direction of positive negative. 
Is it useful? If not we will remove it.

Reviewer: Yes I think it is useful. The convention is also written down in a comment in code: https://github.com/PX4/PX4-Autopilot/blob/96b03040491e727752751c0e0beed87f0966e6d4/src/modules/control_allocator/module.yaml#L492

Reviewer 2: Thanks! Added back.

This section defines the properties of the tilting servos.
These are mapped to specific motors in the [motor geometry for a tiltrotor](../config/actuators.md#motor-geometry-vtol-tiltrotor).

The example below shows the tilt servo setup for the [tiltrotor motor geometry shown above](../config/actuators.md#motor-geometry-vtol-tiltrotor).

![Tilt Servo Geometry Setup Example](../../assets/config/actuators/tilt_servo_geometry_config.png)

The values that can be set are:
- `Tilt servos`: The number of servos (tiltable motors).
- `Angle at min tilt`: [Maximum tilt angle](#tilt-servo-coordinate-system), relative to the z axis (degrees).
- `Angle at max tilt`: [Minimum tilt angle](#tilt-servo-coordinate-system), relative to the z-axis (degrees).
- `Tilt direction`: `Towards front` (Positive X direction) or `Towards right` (Positive Y direction).
- `Use for control`: [Tilt server used for yaw/pitch](#tilt-servos-for-yaw-pitch-control)
  - `None`: Torque control is not used.
  - `Yaw`: Tilt servos used to control yaw.
  - `Pitch`: Tilt servos used to control pitch.
  - `Both Yaw and Pitch`: Tilt servos are used to control both yaw and pitch.

> #### Tilt Servo Coordinate System

Reviewer 2: @bkueng This is how I think tilt servo values work. This is linked to  from section above. The approach is "by example" - which I think removes the ambiguity.

Note that you will need to also ensure that the ESC associated with bidirectional motors is configured appropriately (e.g. 3D mode enabled for DShot ESCs, which can be achieved via [DShot commands](../peripherals/dshot.md#commands)).

### Control Surfaces Geometry

### Control Surface Direction (Conceptual)

![Control Surface Setup Example](../../assets/config/actuators/control_surfaces_geometry.png)

> Example Control Surface Geometry setup for a Tiltrotor VTOL

:::note
It is extremely important that we decouple the concept of control surface, servo and motor tilt directions in control allocation from the actual servo output / connections in the physical vehicle.

That means that the Control Allocation Control Surface & Servo settings are physical vehicle independent. Therefore when we think about the surfaces moving, we imagine the ideal vehicle as shown below that moves actuators in the given direction!
:::

![Control Surface Deflections](../../assets/config/actuators/plane_control_surface_convention.png)

Control surfaces use the following deflection direction convention:

#### Horizontal Control Surface (e.g. Aileron)

> Upwards movement equals Positive deflection

> 'Pitch UP' control input should command positive deflection to both ailerons

Reviewer 2: All of this stuff is very interesting and has nothing to do with configuration of these surfaces. What we need here is an outline of how you set the scaling values. When are they 1? When are they -1. What does 0 mean. Why 0.5?

This section defines the properties of the tilting servos.
These are mapped to specific motors in the [motor geometry for a tiltrotor](../config/actuators.md#motor-geometry-vtol-tiltrotor).

The example below shows the tilt servo setup for the [tiltrotor motor geometry shown above](../config/actuators.md#motor-geometry-vtol-tiltrotor).

![Tilt Servo Geometry Setup Example](../../assets/config/actuators/tilt_servo_geometry_config.png)

The values that can be set are:
- `Tilt servos`: The number of servos (tiltable motors).
- `Angle at min tilt`: [Maximum tilt angle](#tilt-servo-coordinate-system) in degrees, relative to the z axis.
- `Angle at max tilt`: [Minimum tilt angle](#tilt-servo-coordinate-system) in degrees, relative to the z-axis.
- `Tilt direction`: `Towards front` (positive x direction) or `Towards right` (positive y direction).
- `Use for control`: [Tilt servo used for yaw/pitch](#tilt-servos-for-yaw-pitch-control)
  - `None`: Torque control is not used.
  - `Yaw`: Tilt servos used to control yaw.
  - `Pitch`: Tilt servos used to control pitch.
  - `Both Yaw and Pitch`: Tilt servos are used to control both yaw and pitch.

> #### Tilt Servo Coordinate System

Reviewer 2: Rewrote this again after improving the diagram. I understand this now.

![Tilt Axis](../../assets/config/actuators/tilt_axis.png)

The `Angle at min tilt` and `Angle at max tilt` indicate the range of movement for the tilt servo.
The minimum tilt is the smaller _numerical value_ (not absolute) of the two angles.

If the max/min tilt vectors are **P<sub>0</sub>** and **P<sub>1</sub>** as shown above, both tilt angles are positive but **θ<sub>0</sub>** is smaller:
- `Angle at min tilt` = **θ<sub>0</sub>**
- `Angle at max tilt` = **θ<sub>1</sub>**

:::note
If the diagram was mirrored so that **P<sub>0</sub>** and **P<sub>1</sub>** were tilting into the -x, -y quadrant, then both the tilt angles would be negative.
Because **θ<sub>1</sub>** would more negative (smaller) than **θ<sub>0</sub>**, it would be the `Angle at min tilt`.

Similarly, a servo that moves:
- between the upright and forward postions would have `min=0` and `max=90`. 
- symmetrically 45 degrees around the upright position would have `min=-45` and `max=45`
- between the upright and backward positions would have `min=-90` and `max=0`.
:::

The `Tilt direction` indicates whether the servo tilts in the plane towards the `Front` or `Right` of the vehicle.

> On the diagram this would be represented by **α** that can only take values of 0 (front) or 90 (right).

Reviewer 2: The values might be wrong. My assumption is that tilt direction is 0 pointing forward, increasing positive in rotation towards the right.

Reviewer: Sounds correct to me.

* Example : To roll in positive direction (banking to the right), the 'Right Aileron' needs a positive deflection.

#### Vertical Control Surface (e.g. Rudder)
> Rightwards movement equals Positive Deflection

'Yaw RIGHT' control input should command positive deflection.

* Example : To yaw in positive direction (clockwise rotation), conventional rudder airplane needs a positive deflection.

#### Mixed Control Surface (e.g. V-Tail)
> Upwards movement equals Positive deflection

'Pitch UP' control input should command positive deflection to both V-tails

* Example : To pitch in positive direction (pitching upwards),  'Both V-tail surfaces' need a positive deflection.

> Rightwards movement equals Positive Deflection
'Yaw RIGHT' control input should command positive deflection to both V-tails.

* Example : Having a V or A shaped tail in the back, positive deflection in Rightwards direction will Yaw the vehicle to the right.

### Motor Tilt Servos (Conceptual)
![Tilt Servo Geometry Setup Example](../../assets/config/actuators/tilt_servo_geometry_config.png)

> Example Tilt Servo Geometry setup for a Tiltrotor VTOL

> Similar to the Control Surfaces, Servos can also rotate the motors themselves! This is often the case in a VTOL like [Tiltrotors](/en/frames_vtol/tiltrotor.md). 

Reviewer 3: How is this similar?

> Rightwards movement equals Positive Deflection
'Yaw RIGHT' control input should command positive deflection to both V-tails.

* Example : Having a V or A shaped tail in the back, positive deflection in Rightwards direction will Yaw the vehicle to the right.

### Motor Tilt Servos (Conceptual)
![Tilt Servo Geometry Setup Example](../../assets/config/actuators/tilt_servo_geometry_config.png)

> Example Tilt Servo Geometry setup for a Tiltrotor VTOL

Similar to the Control Surfaces, Servos can also rotate the motors themselves! This is often the case in a VTOL like [Tiltrotors](/en/frames_vtol/tiltrotor.md). 

### Motor Tilt Servos
![Tilt Axis](../../assets/config/actuators/tilt_axis.png)

Tilt servos are configured as follows:
- The reference direction is upwards (negative Z direction).
- Tilt direction: **Towards Front** means the servo tilts towards positive X direction, whereas **Towards Right** means towards positive Y direction.
- Minimum and maximum tilt angles: specify the physical limits in degrees of the tilt at minimum control and maximum respectively.

>   An angle of 0° means to point upwards, then increases towards the tilt direction.

Reviewer 3: Why did you remove this?

Reviewer 2: Either way I don't understand the way this works/coordinate system. 

My understanding of tiltrotors comes from the eflight convergence. This has a rotor that points up to hover and forward to fly forward. My understanding is that this would be set to "Towards Front". It points either forward or straight up but it isn't clear to me what the min and max tilt angle will be - I know one will be 0, but not which position, and also whether the other value will be 90 or -90.
Also, what does it mean if the motor goes to the right? Does that mean you can't use it as a pusher?

Reviewer 3: There's a lot of degrees of freedom here, hence it's complicated. You can think of the tilt rotating within a vertical plane, with the 'yaw' of the plane given by the tilt direction (front or right in that case).
Then within the plan you have a starting angle and an end angle. You can think of a clock: 12=upwards=0 deg=always aligned with Z axis, then the angle increases clockwise (if you were to look at the plane from the right side).
Not sure if that helps or makes it even more confusing.

Reviewer 2: For the eflight I can kind of see this. The min angle is 0 when in hover, and the max angle is 90 and the tilt is forward. From the right I can see the motor moving from its spindle being pointed up to being forwards, a move from 12 oclock (0) to 3oclock (90). 

But if the motor was moving right then I would see the props coming from invisible to visible and facing me. I guess the move is still 0 to 90. But what use is it? When facing me it is pulling the VTOL "plane" sideways?

IF I had an eflight that used a prop like a pusher below the vehicle and then moved forward, the angle would change from -180 to 90 anticlockwise (from right). 

Nup. I guess I don't get it :-(. I do know this does not pass the "idiot gets it test"

> * Tilting **Front** means the servo tilts towards Front (Positive X direction).

Reviewer 3: ```suggestion
* Tilting towards **Front** means the servo tilts towards Front (Positive X direction).
```

For VTOL tiltrotor vehicles, it will also include the number and properties of [tilt servos](#motor-tilt-servo-geometry)

:::note

> The UI shows the configured motor positions, customized to your vehicle (Currently only for Multicopter)

Reviewer 2: This change is wrong. This is about the whole geometry section not the multicopter image. 

Or by example, if you have a vehicle with motors that you can't customise the geometry on, they won't be displayed."

I will revert.

Reviewer 2: Actually, I'm going to merge the two points :-)

- `Position Y`: [Y-position](#motor-position-coordinate-system), in metres.
- `Position Z`: [Z-position](#motor-position-coordinate-system), in metres.
- (Advanced) `Direction CCW`: Checkbox to indicate motor spins counter-clockwise (uncheck for clockwise).
- (Advanced) `Bidirectional`: Checkbox to indicate motor is [bidirectional](#bidirectional-motors) 
- (Advanced) `Slew Rate`: ?

The X, Y, Z positions are in [FRD coordinate frame, relative to the centre of gravity](#motor-position-coordinate-system).

#### Motor Geometry: VTOL Quadrotor Tailsitter

VTOL Quadrotor is a tailsitter with 4 motors like a quadcopter as shown below.

![VTOL Quadrotor Tailsitter](../../assets/airframes/types/VTOLQuadRotorTailSitter.svg)
![VTOL Quadrotor Tailsitter Real Example](../../assets/airframes/vtol/skypull/skypull_sp1.jpg)

The default motor geometry is shown below. Motors have the same configuration fields as for the [multicopter geometry](#motor-geometry-multicopter).

![Geometry motor: tailsitter vtol](../../assets/config/actuators/qgc_geometry_tailsitter_motors.png)

#### Motor Geometry: VTOL Tiltrotor

The default VTOL tiltrotor motor geometry is shown below.

> ![Tiltrotor VTOL Airframe](../../assets/airframes/types/VTOLTiltRotor.svg)

Reviewer 2: That image is a hexarotor tiltrotor. 

It is just too big. Ditto with the quad tailstitter. 

FYI only. My point - worth rendering this to see what it looks like. To be clear, I think having an image is a good idea, but one that pushes the content a page down is annoying for readers.

I'll see if I can constrain these.

Author: I agree. The Airfarme SVGs are quite confusing! Will create PR for that as well

Author: PR added here: https://github.com/PX4/PX4-Autopilot/pull/19686

## Actuator Testing

### Identify & Assign Motors

![](../../assets/config/actuators/identify_motors_button.png)

> QGC Supports a feature of identifying the motors by sending the motor command to different output pins and letting you click on the motor to map the Motor to each Output ports.

Reviewer 3: ```suggestion
QGC Supports a feature of identifying the motors by sending the motor command to different outputs and letting you click on the motor to map the Motor to each Output ports.
```

To use this mode:

> 1. Setup the Motors Geometry correctly to include all the motors in your vehicle you want to configure

Reviewer 3: Add: 2. select the output tab where you want to assign the motors, if not assigned already

Reviewer 2: I might revisit this a bit later, but for now adding the point.,

```suggestion
1. Setup the Motors Geometry correctly to include all the motors in your vehicle you want to configure
1. Select the Output tab where you want to assign the motors (if not assigned already)
```

2. Click the 'Identify & Assign Motors' Button

> 3. Each motor will start spinning. Select the motor that spun up in the GUI.

Reviewer 3: ```suggestion
3. One motor will start spinning. Select the motor that spun up in the GUI.
```

:::

More detailed setup directions are described below.

#### Motor setup & testing
:::note
Always remove the propellers from the motor before you do the Actuator testing for Safety!!
:::

1. Pull the Motor slider down, and it will snap in lower end, where it will command **disarmed** value to the motors, verify that it doesn't spin
   1. For PWM output, adjust the Disarmed value in case the motor spins at the disarmed command
2. Slowly move the Motor slider up, and it will snap in at the '**minimum** command'. Verify that the motor is spinning at the minimum speed.
   1. For **PWM output, adjust the minimum output** value in the 'Actuator Outputs' tab on the right such that the motors barely spin
    ![PWM Minimum Output](../../assets/config/actuators/pwm_minimum_output.png)
   :::note
   For DShot output, this is not required
   :::
3. Move the slider up even more, and verify that the Motor is spinning in the correct direction and that it would give a positive thrust in the expected direction
   1. The expected thrust direction can vary by vehicle type, for example in multicopters the thrust should always point upwards

>    2. For Fixed Wings, the thrust should point to forward direction

Reviewer 3: Isn't that too basic?

Author: What do you mean by basic?

Reviewer 2: Beat means "simple". Even the most ignorant user will know that thrust is forwards on a fixed wing vehicle.

   1. If it is not aligned, you can set the **Trim** value for each Control Surface.
   2. After setting a new trim, manually click and place the slider for the control surface you are testing in the middle position, to check if it is in the neutral position.
   3. Note that as mentioned above as a safety feature of slider, unless you actually touch the slider, the actuator command doesn't get sent. Therefore **you must manually click and put the slider in the middle position to test the trim**
   4. Or, an easier way to test without using the slider would be to set the [`COM_PREARM_MODE`](../advanced_config/parameter_reference.md#COM_PREARM_MODE) parameter to `Always`.
      1. Which will enable the control of Servos even when the vehicle is disarmed, and will constantly be applying the Trim setting to the Control Surfaces
      2. Here you can try setting different values for the Trim and check the alignment, and then settle on the value you are happy with.

#### Tilt Servo setup
![Tilt Servo Setup](../../assets/config/actuators/tilt_servo_setup.png)

> Setup of Tilt Servo is similar to Control Surface, as both of them uses the Servo hardware (in most cases)

> 1. **Set the 'Disarmed' value for Control Surface outputs to '1500'** in case of using a PWM output

Reviewer 3: This depends on the vtol, some might want to use 1000 or 2000.

Reviewer 2: Thanks @bkueng. I changed this to as shown below. If this is not OK let me know (since this text may be moved).

```suggestion
1. **Set the 'Disarmed' value for control surface outputs. The value depends on the vehicle.
```

Are there any particular "considerations" for this? I mean for a _failsafe_ value on an FW vehicle it is not uncommon to set the value such that the vehicle does a slow glide or circle to land. For an MC motor you want a _disarmed_ value to not power a motor. This is kind of FMI, but perhaps might be useful to provide additional hints above.

:::note
After you toggle the **Enable sliders** switch, actuators/motors won't do anything until the corresponding slider is _moved_.
This is a safety feature to prevent sudden motor movements after switch is enabled.
:::

Sliders can be used to verify the following:

> 1. Actuators (Motors, Control surfaces, etc.) are assigned to the expected output.

Reviewer 2: Nice!

1. Motors don't spin when at the `disarmed` PWM output value
1. Motors barely spin at the `minimum` PWM output value.
1. Motors give **positive thrust** in the expected direction

> 1. Control Surfaces are in the correct idle position for `disarmed` output value

Reviewer 2: Yes, good consistency change. This is what I was trying to capture in my changes - we're slotting to the point that feeds the specified field value :-).
